miden-lib/asm/sat/epilogue.masm

```diff
@@ -0,0 +1,218 @@
+end
+
+#! Computes the hash of a created note with index i. This is computed as follows:
+#! - we define, recipient =
+#!       hash(hash(hash(serial_num, [0; 4]), script_hash), input_hash)
+#! - we then compute the created note hash as:
+#!       hash(recipient, vault_hash)
+#!
+#! Stack: [note_data_ptr]
+#! Output: [CREATED_NOTE_HASH]
+#!
+#! - note_data_ptr is a pointer to the data section of the created note.
+#! - CREATED_NOTE_HASH is the hash of the created note with index i.
+proc.compute_created_note_hash
+    # pad capacity elements of hasher
+    padw
+
+    # insert created note recipient into the first four elements of the hasher rate
+    padw dup.8 push.CREATED_NOTE_RECIPIENT_OFFSET add mem_loadw
+
+    # populate the last four elements of the hasher rate with the created note vault hash
+    dup.8 exec.compute_created_note_vault_hash
+
+    # compute created note hash and extract digest
+    hperm dropw swapw dropw
+
+    # save created note hash to memory
+    movup.4 mem_storew
+end
+
+#! Computes a commitment to the created notes. This is computed as a sequential hash of
+#! (note_hash, note_metadata) tuples.
+#!
+#! Stack: []
+#! Output: [CREATED_NOTES_COMMITMENT]
+#!
+#! - CREATED_NOTES_COMMITMENT is the commitment of the created notes.
+proc.process_created_notes
+    # get the number of created notes from memory
+    push.NUM_CREATED_NOTES_PTR mem_load
+
+    # calculate the address at which we should stop looping
+    exec.utils::get_created_note_data_ptr
+
+    # compute pointer for first address
+    push.0 exec.utils::get_created_note_data_ptr
+
+    # prepare stack for hashing
+    padw padw padw
+
+    # check if the number of created notes is greater then 0. Conditional for the while loop.
+    dup.13 dup.13 neq
+
+    # loop and hash created notes
+    while.true
+        # compute created note hash (this also computes created not vault hash)
+        dup.12 exec.compute_created_note_hash
+
+        # drop created note hash from stack
+        dropw
+
+        # permute over (note_hash, note_metadata)
+        mem_stream
+
+        # increment created note pointer and check if we should loop again
+        movup.12 push.CREATED_NOTE_HASHING_MEM_DIFF add dup movdn.13 dup.14 neq
+    end
+
+    # extract digest from hasher rate elements (h_0, ..., h_3)
+    dropw swapw dropw
+
+    # drop accessory variables from stack
+    movup.4 drop
+    movup.4 drop
+end
+
+# FINALIZE NONCE PROCEDURE
+# =================================================================================================
+
+#! Asserts that the final account nonce is greater than the initial account nonce.
+#!
+#! Stack: []
+#! Output: []
+proc.finalize_nonce
+    # get initial nonce from memory
+    exec.utils::get_initial_account_nonce
+
+    # get current nonce from memory
+    exec.utils::get_account_nonce
+
+    # assert that initial nonce is less than current nonce
+    lt assert
+end
```

Comment:
I'm not sure we should break this out into a separate procedure - it seems simple enough that it can be inlined on line 204 below. But not a strong preference from my side.